File: Boj.단계별/14_백트래킹/javascript/2580.js
let input = require('fs').readFileSync('/dev/stdin').toString().split('\n');
let target = [];
let checkBool;
const isEmptyArray = (arr) => {
    if (Array.isArray(arr) && arr.length === 0){
        return true;
    }
    return false;
}
for (let i = 0 ; i<9 ; i++){
  input[i] = input[i].split(' ').map(Number);
}

const check = (r,c,num) => {
    for (let i = 0 ; i<9 ; i++){
        if (input[r][i] === num) return false;
        if (input[i][c] === num) return false;
    }
    let [x,y] = [Math.floor(r/3)*3, Math.floor(c/3)*3];
    for (let i = 0 ; i<3 ; i++){
        for (let j = 0 ; j<3 ; j++){
            if( input[x+i][y+j] === num) return false;
        }
    }
    return true;
}
const solution = (cnt) => {
    if (cnt == target.length){
        for (let i = 0 ; i<9 ; i++){
            console.log(input[i].join(" "));
        }
        checkBool = true;
        return false;
    }
    let [r,c] = [target[cnt][0],target[cnt][1]];
    for (let i = 1 ; i<=9 ; i++){
        if (checkBool){
            return false;
        }
        if (check(r,c,i)){
            input[r][c] = i;
            solution(cnt+1);
            input[r][c] = 0;
        }
    }
    
}
for (let i = 0 ;i<9 ;i++){
    for (let j = 0 ; j<9 ; j++){
        if (input[i][j] == 0){
            target.push([i,j])
        }
    }
}
solution(0);

File: Backtracking/python/2580.py
import sys
input = sys.stdin.readline

nums = [list(map(int,input().split())) for _ in range(9)]
target = [[i,j] for i in range(9) for j in range(9) if nums[i][j] == 0]


def check(r,c,num):
  for i in range(9):
    if nums[r][i] == num : return False
    if nums[i][c] == num : return False
  x,y = (r//3)*3, (c//3)*3
  for i in range(3):
    for j in range(3):
      if nums[x+i][y+j] == num : return False
  return True;
def solution(cnt):
  if cnt == len(target):
    for num in nums:
      sys.stdout.write(" ".join(map(str,num))+"\n")
    exit()
  r,c = target[cnt][0],target[cnt][1]
  for i in range(1,10):
    if (check(r,c,i)):
      nums[r][c] = i
      solution(cnt+1)
      nums[r][c] = 0
solution(0)